woocommerce/woocommerce-admin
# Match server-side CSV export format to client-side

Fixes #2980.
Fixes https://github.com/woocommerce/woocommerce-admin/issues/3023

This PR seeks to align the server-side and client-side report export formats by:

* Introducing a `ExportableInterface` which requires `get_export_columns()` and `prepare_item_for_export()` methods to be defined on Controllers
* Allows and escapes commas and double quotes in CSV values
* Uses `WP_REST_Server::response_to_data()` to include embedded objects in responses

One kludge I'm not fond of is the `get_export_items()` check - it's there so we can shape the response of `/revenue/stats` into a single array of values. I'm open to ideas on how to better implement that.

### Detailed test instructions:

For each report:

- Use date range and filters to get a single-page result set
- Download the report (should be in-browser)
- Use date range and filters to get a multi-page result set
- Download the report (should be server-side, emailed)
- Compare CSV files, they should match



### Changelog Note:



Fix: report export format when generated server-side.
